app/controllers/publishers/vacancies/job_applications/messages_controller.rb

```diff
@@ -2,26 +2,37 @@ class Publishers::Vacancies::JobApplications::MessagesController < Publishers::V
   before_action :set_job_application
 
   def create
-    conversation = find_or_create_conversation
-    message = conversation.messages.build(message_form_params.merge(sender: current_publisher))
+    message_form = Publishers::JobApplication::MessagesForm.new(message_form_params)
 
-    if message.valid?
-      conversation.save! unless conversation.persisted?
-      message.save!
-      redirect_to organisation_job_job_application_path(id: @job_application.id, tab: "messages"), success: t(".success")
+    if message_form.valid?
+      conversation = find_or_create_conversation
+      Message.create!(content: message_form.content, sender: current_publisher, conversation: conversation)
+      redirect_to organisation_job_job_application_path(@vacancy.id, @job_application.id, tab: "messages"), success: t(".success")
     else
-      redirect_to organisation_job_job_application_path(id: @job_application.id, tab: "messages"), warning: t(".failure")
+      # Set params to simulate the show action with show_form=true and tab=messages
+      params[:tab] = "messages"
+      params[:show_form] = "true"
+
+      conversation = @job_application.conversations.first
+      @messages = if conversation&.messages
+                    conversation.messages.order(created_at: :desc)
+                  else
+                    []
+                  end
+      @message_form = message_form
+
+      render "publishers/vacancies/job_applications/show", status: :unprocessable_entity
     end
   end
 
   private
 
   def find_or_create_conversation
     @job_application.conversations.first ||
-      @job_application.conversations.build(title: Conversation.default_title_for(@job_application))
+      @job_application.conversations.create!(title: Conversation.default_title_for(@job_application))
```

Comment:
this change creates a bug - it creates a conversation even if the message isn't valid.